From fb5a372cb867a1d95b59283def91ac2b77689f03 Mon Sep 17 00:00:00 2001
From: Viren070 <viren070@protonmail.com>
Date: Tue, 27 Oct 2026 20:45:25 +0000
Subject: [PATCH] ass_render: render events on helper threads

ass_set_threads() sets how many threads render a frame's events, by
default one per processor up to 8. The calling thread renders too, and
wakes helpers only for frames predicted to take more than 16 ms, as
few as are predicted to bring the frame under that: each event is
expected to cost what it did last frame, and a new one what new events
of its layer and style have lately. Costly events go first. Frames whose
threads keep waiting for each other's cache items use fewer.
---
 libass/Makefile_library.am |   1 +
 libass/ass.h               |  12 +
 libass/ass_cache.c         |   8 +-
 libass/ass_cache.h         |   1 +
 libass/ass_render.c        |  27 ++-
 libass/ass_render.h        |  17 ++
 libass/ass_render_api.c    |  12 +
 libass/ass_render_pool.c   | 450 +++++++++++++++++++++++++++++++++++++
 libass/ass_utils.h         |   8 +
 libass/libass.sym          |   1 +
 libass/meson.build         |   1 +
 11 files changed, 526 insertions(+), 12 deletions(-)
 create mode 100644 libass/ass_render_pool.c

diff --git a/libass/Makefile_library.am b/libass/Makefile_library.am
index ec7bff9..6c85c48 100644
--- a/libass/Makefile_library.am
+++ b/libass/Makefile_library.am
@@ -23,6 +23,7 @@ libass_libass_internal_la_SOURCES = \
     libass/ass_bitmap.h libass/ass_bitmap.c libass/ass_blur.c \
     libass/ass_rasterizer.h libass/ass_rasterizer.c \
     libass/ass_render.h libass/ass_render.c libass/ass_render_api.c \
+    libass/ass_render_pool.c \
     libass/ass_bitmap_engine.h libass/ass_bitmap_engine.c \
     libass/ass_arabic_charmap.h libass/ass_arabic_charmap.c \
     libass/c/rasterizer_template.h libass/c/c_rasterizer.c \
diff --git a/libass/ass.h b/libass/ass.h
index 81ecec1..b2804c9 100644
--- a/libass/ass.h
+++ b/libass/ass.h
@@ -595,6 +595,18 @@ void ass_set_selective_style_override(ASS_Renderer *priv, ASS_Style *style);
 ASS_DEPRECATED("it does nothing")
 int ass_fonts_update(ASS_Renderer *priv);
 
+/**
+ * \brief Set how many threads render events: the one calling
+ * ass_render_frame(), and helpers it starts when a frame is costly enough.
+ * The message callback may then be called from the helpers, one message at
+ * a time.
+ * \param priv renderer handle
+ * \param threads 1 for the calling thread alone, 0 for the default: the
+ * number of processors, at most 8
+ * \return the number of threads that will be used
+ */
+int ass_set_threads(ASS_Renderer *priv, int threads);
+
 /**
  * \brief Set hard cache limits.  Do not set, or set to zero, for reasonable
  * defaults.  The default bitmap cache size is for a 4K frame and shrinks with
diff --git a/libass/ass_cache.c b/libass/ass_cache.c
index 0948b61..40fde71 100644
--- a/libass/ass_cache.c
+++ b/libass/ass_cache.c
@@ -490,17 +490,19 @@ static CacheItem *find_item(CacheItem *item, CacheItem *end, ass_hashcode hash,
     return NULL;
 }
 
-static void wait_until_ready(Cache *cache, CacheItem *item)
+static void wait_until_ready(Cache *cache, CacheClient *client, CacheItem *item)
 {
     if (atomic_load_explicit(&item->state, memory_order_acquire) == ITEM_READY)
         return;
 
+    int64_t start = ass_now_us();
     pthread_mutex_lock(&cache->wait_lock);
     unsigned state = ITEM_CONSTRUCTING;
     atomic_compare_exchange_strong(&item->state, &state, ITEM_AWAITED);
     while (atomic_load_explicit(&item->state, memory_order_acquire) != ITEM_READY)
         pthread_cond_wait(&cache->wait_cond, &cache->wait_lock);
     pthread_mutex_unlock(&cache->wait_lock);
+    client->wait_us += ass_now_us() - start;
 }
 
 static void mark_ready(Cache *cache, CacheItem *item)
@@ -539,7 +541,7 @@ void *ass_cache_get(Cache *cache, CacheClient *client, void *key, void *priv)
     if (item) {
         desc->key_move_func(NULL, key);
         use_item(cache, client, item);
-        wait_until_ready(cache, item);
+        wait_until_ready(cache, client, item);
         return (char *) item + CACHE_ITEM_SIZE;
     }
 
@@ -574,7 +576,7 @@ void *ass_cache_get(Cache *cache, CacheClient *client, void *key, void *priv)
             desc->key_destruct_func(new_key);
             free(item);
             use_item(cache, client, other);
-            wait_until_ready(cache, other);
+            wait_until_ready(cache, client, other);
             return (char *) other + CACHE_ITEM_SIZE;
         }
         searched = head;
diff --git a/libass/ass_cache.h b/libass/ass_cache.h
index 3730365..b6ab22f 100644
--- a/libass/ass_cache.h
+++ b/libass/ass_cache.h
@@ -113,6 +113,7 @@ typedef struct
 // ass_cache_flush() moves to the back of their caches' eviction order.
 typedef struct {
     struct cache_item *recent;
+    int64_t wait_us;    // spent waiting for items other threads were constructing
 } CacheClient;
 
 // Lookups may run on several threads at once, each with its own client.
diff --git a/libass/ass_render.c b/libass/ass_render.c
index f4ebd47..20d6fcb 100644
--- a/libass/ass_render.c
+++ b/libass/ass_render.c
@@ -76,7 +76,7 @@ static void text_info_done(TextInfo* text_info)
     free(text_info->combined_bitmaps);
 }
 
-static bool render_context_init(RenderContext *state, ASS_Renderer *priv)
+bool ass_render_context_init(RenderContext *state, ASS_Renderer *priv)
 {
     state->renderer = priv;
 
@@ -90,7 +90,7 @@ static bool render_context_init(RenderContext *state, ASS_Renderer *priv)
     return ass_rasterizer_init(&priv->engine, &state->rasterizer, RASTERIZER_PRECISION);
 }
 
-static void render_context_done(RenderContext *state)
+void ass_render_context_done(RenderContext *state)
 {
     ass_rasterizer_done(&state->rasterizer);
 
@@ -151,7 +151,7 @@ ASS_Renderer *ass_renderer_init(ASS_Library *library)
 
     priv->cache.glyph_max = GLYPH_CACHE_MAX;
 
-    if (!render_context_init(&priv->state, priv))
+    if (!ass_render_context_init(&priv->state, priv))
         goto fail;
 
     priv->user_override_style.Name = "OverrideStyle"; // name insignificant
@@ -161,6 +161,7 @@ ASS_Renderer *ass_renderer_init(ASS_Library *library)
 
     ass_shaper_info(library);
     priv->settings.shaper = ASS_SHAPING_COMPLEX;
+    priv->settings.threads = ass_default_threads();
 
     ass_msg(library, MSGL_V, "Initialized");
 
@@ -178,6 +179,8 @@ void ass_renderer_done(ASS_Renderer *render_priv)
     if (!render_priv)
         return;
 
+    ass_render_pool_free(render_priv->pool);
+
     ass_frame_unref(render_priv->images_root);
     ass_frame_unref(render_priv->prev_images_root);
 
@@ -195,7 +198,7 @@ void ass_renderer_done(ASS_Renderer *render_priv)
         FT_Done_FreeType(render_priv->ftlibrary);
     free(render_priv->eimg);
 
-    render_context_done(&render_priv->state);
+    ass_render_context_done(&render_priv->state);
 
     free(render_priv->settings.default_font);
     free(render_priv->settings.default_family);
@@ -2804,7 +2807,7 @@ static void add_background(RenderContext *state, EventImages *event_images)
  * \param event_images struct containing resulting images, will also be initialized
  * Process event, appending resulting ASS_Image's to images_root.
  */
-static bool
+bool
 ass_render_event(RenderContext *state, ASS_Event *event,
                  EventImages *event_images)
 {
@@ -3044,7 +3047,7 @@ static void check_cache_limits(ASS_Renderer *priv, CacheStore *cache)
     ass_cache_cut(cache->shape_cache, SHAPE_CACHE_MAX_SIZE);
 }
 
-static void setup_shaper(ASS_Shaper *shaper, ASS_Renderer *render_priv)
+void ass_setup_shaper(ASS_Shaper *shaper, ASS_Renderer *render_priv)
 {
     ASS_Track *track = render_priv->track;
 
@@ -3090,7 +3093,7 @@ ass_start_frame(ASS_Renderer *render_priv, ASS_Track *track,
             render_priv->fontselect, render_priv->num_emfonts);
     }
 
-    setup_shaper(render_priv->state.shaper, render_priv);
+    ass_setup_shaper(render_priv->state.shaper, render_priv);
 
     // PAR correction
     double par = render_priv->settings.par;
@@ -3408,11 +3411,17 @@ ASS_Image *ass_render_frame(ASS_Renderer *priv, ASS_Track *track,
                     realloc(priv->eimg,
                             priv->eimg_size * sizeof(EventImages));
             }
-            if (ass_render_event(&priv->state, event, priv->eimg + cnt))
-                cnt++;
+            priv->eimg[cnt++] = (EventImages) { .event = event };
         }
     }
 
+    ass_render_events(priv, cnt);
+    int rendered = 0;
+    for (int i = 0; i < cnt; i++)
+        if (priv->eimg[i].rendered)
+            priv->eimg[rendered++] = priv->eimg[i];
+    cnt = rendered;
+
     // sort by layer
     if (cnt > 0)
         qsort(priv->eimg, cnt, sizeof(EventImages), cmp_event_layer);
diff --git a/libass/ass_render.h b/libass/ass_render.h
index b831553..a226ac9 100644
--- a/libass/ass_render.h
+++ b/libass/ass_render.h
@@ -80,6 +80,8 @@ typedef struct {
 
     char *default_font;
     char *default_family;
+
+    int threads;    // that render events, the calling one included
 } ASS_Settings;
 
 // a rendered event
@@ -89,6 +91,8 @@ typedef struct {
     int detect_collisions;
     int shift_direction;
     ASS_Event *event;
+    bool rendered;
+    uint32_t cost_us;   // how long rendering took, or is expected to take
 } EventImages;
 
 typedef enum {
@@ -313,6 +317,8 @@ typedef struct {
     size_t composite_max_size;
 } CacheStore;
 
+typedef struct render_pool RenderPool;
+
 struct ass_renderer {
     ASS_Library *library;
     FT_Library ftlibrary;
@@ -339,6 +345,7 @@ struct ass_renderer {
 
     RenderContext state;
     CacheStore cache;
+    RenderPool *pool;
 
     BitmapEngine engine;
 
@@ -358,6 +365,16 @@ typedef struct {
 } Rect;
 
 void ass_reset_render_context(RenderContext *state, ASS_Style *style);
+bool ass_render_context_init(RenderContext *state, ASS_Renderer *priv);
+void ass_render_context_done(RenderContext *state);
+void ass_setup_shaper(ASS_Shaper *shaper, ASS_Renderer *render_priv);
+bool ass_render_event(RenderContext *state, ASS_Event *event,
+                      EventImages *event_images);
+
+int ass_default_threads(void);
+// Renders the frame's events in eimg, on helper threads too when it pays.
+void ass_render_events(ASS_Renderer *priv, int count);
+void ass_render_pool_free(RenderPool *pool);
 void ass_frame_ref(ASS_Image *img);
 void ass_frame_unref(ASS_Image *img);
 ASS_Vector ass_layout_res(ASS_Renderer *render_priv);
diff --git a/libass/ass_render_api.c b/libass/ass_render_api.c
index 6e809f3..bccdcf6 100644
--- a/libass/ass_render_api.c
+++ b/libass/ass_render_api.c
@@ -190,6 +190,18 @@ int ass_fonts_update(ASS_Renderer *render_priv)
     return 1;
 }
 
+int ass_set_threads(ASS_Renderer *priv, int threads)
+{
+    if (threads <= 0)
+        threads = ass_default_threads();
+    if (threads != priv->settings.threads) {
+        ass_render_pool_free(priv->pool);
+        priv->pool = NULL;
+        priv->settings.threads = threads;
+    }
+    return threads;
+}
+
 void ass_set_cache_limits(ASS_Renderer *render_priv, int glyph_max,
                           int bitmap_max)
 {
diff --git a/libass/ass_render_pool.c b/libass/ass_render_pool.c
new file mode 100644
index 0000000..843b07a
--- /dev/null
+++ b/libass/ass_render_pool.c
@@ -0,0 +1,450 @@
+/*
+ * Copyright (C) 2026 Viren070
+ *
+ * This file is part of libass.
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include "config.h"
+#include "ass_compat.h"
+
+#include <pthread.h>
+#include <stdatomic.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "ass_render.h"
+
+// A frame's events render on the calling thread, which wakes as few helper
+// threads as are predicted to bring the frame under TARGET_US: more only cost
+// more, as threads on the same cores slow each other down. An event is
+// expected to cost what it did last frame, and a new one what new events of
+// its layer and style have lately. Costly events go first, costliest first,
+// so that no long one starts last. Threads that keep waiting for each other's
+// cache items, as when their events need the same new glyphs, are worth less
+// than they cost: after two such frames in a row, frames use half as many
+// threads, and one more again after every CALM_FRAMES frames without.
+
+#define MAX_DEFAULT_THREADS 8
+#define TARGET_US 16000
+#define NEW_EVENT_US TARGET_US  // a new event's expected cost, with nothing to go by
+#define HEAVY_EVENT_US 500      // events this costly are taken one at a time
+#define NEW_COST_SLOTS 256      // a power of two
+#define CALM_FRAMES 8
+
+typedef struct {
+    int index;          // in the track's events
+    int read_order;     // tells the event apart after the track changed
+    uint32_t cost_us;
+} EventCost;
+
+// What new events of a layer and style have lately cost: a running mean that
+// each new one moves a quarter of the way to its own cost.
+typedef struct {
+    bool used;
+    int layer, style;
+    uint32_t cost_us;
+} NewEventCost;
+
+typedef struct {
+    uint32_t cost_us;   // expected
+    int eimg;
+} PlannedEvent;
+
+typedef struct {
+    pthread_t thread;
+    RenderContext context;
+    struct render_pool *pool;
+} Helper;
+
+struct render_pool {
+    ASS_Renderer *renderer;
+
+    pthread_mutex_t lock;
+    pthread_cond_t work;    // idle helpers wait here
+    pthread_cond_t done;    // the calling thread waits here for busy helpers
+    Helper *helpers;
+    int n_helpers;
+    bool started;
+    int wanted;             // helpers still to join the current frame
+    int busy;               // helpers working on it
+    int joined;             // helpers that worked on it
+    bool quit;
+
+    int max_threads;        // a frame may use, lowered while threads wait on each other
+    int contended, calm;    // frames in a row with and without that
+
+    // the current frame: its events in the order to render them, and which
+    // are new, by index in eimg
+    PlannedEvent *plan;
+    bool *fresh;
+    int count, n_heavy, batch;
+    atomic_int next;
+
+    // last frame's costs, in event order
+    ASS_Track *costs_track;
+    EventCost *costs;
+    int n_costs;
+    NewEventCost new_costs[NEW_COST_SLOTS];
+
+    int max_events;
+};
+
+int ass_default_threads(void)
+{
+    long cores = sysconf(_SC_NPROCESSORS_ONLN);
+    return FFMINMAX(cores, 1, MAX_DEFAULT_THREADS);
+}
+
+static void render_event(RenderContext *state, EventImages *event_images)
+{
+    int64_t start = ass_now_us(), waited = state->cache_client.wait_us;
+    event_images->rendered = ass_render_event(state, event_images->event, event_images);
+    int64_t cost = ass_now_us() - start - (state->cache_client.wait_us - waited);
+    event_images->cost_us = FFMINMAX(cost, 0, UINT32_MAX);
+}
+
+static void wake_helpers(RenderPool *pool, int threads);
+
+// The threads it takes to render the events not yet taken within TARGET_US
+static int threads_for_rest(RenderPool *pool)
+{
+    int64_t rest = 0;
+    for (int i = atomic_load_explicit(&pool->next, memory_order_relaxed); i < pool->count; i++)
+        rest += pool->plan[i].cost_us;
+    return FFMIN((rest + TARGET_US - 1) / TARGET_US, pool->max_threads);
+}
+
+// Renders the frame's events until none are left. Past check_at, which only
+// the calling thread sets, helpers are woken if what remains needs them.
+static void run_queue(RenderPool *pool, RenderContext *state, int64_t check_at)
+{
+    EventImages *eimg = pool->renderer->eimg;
+    for (;;) {
+        int i = atomic_load_explicit(&pool->next, memory_order_relaxed);
+        int n = i < pool->n_heavy ? 1 : pool->batch;
+        i = atomic_fetch_add_explicit(&pool->next, n, memory_order_relaxed);
+        if (i >= pool->count)
+            return;
+        n = FFMIN(n, pool->count - i);
+        for (int j = i; j < i + n; j++)
+            render_event(state, eimg + pool->plan[j].eimg);
+
+        if (check_at && ass_now_us() >= check_at) {
+            wake_helpers(pool, threads_for_rest(pool));
+            check_at = 0;
+        }
+    }
+}
+
+static void *run_helper(void *arg)
+{
+    Helper *helper = arg;
+    RenderPool *pool = helper->pool;
+
+    pthread_mutex_lock(&pool->lock);
+    for (;;) {
+        while (!pool->quit && !pool->wanted)
+            pthread_cond_wait(&pool->work, &pool->lock);
+        if (pool->quit)
+            break;
+        pool->wanted--;
+        pool->busy++;
+        pool->joined++;
+        pthread_mutex_unlock(&pool->lock);
+
+        ass_setup_shaper(helper->context.shaper, pool->renderer);
+        run_queue(pool, &helper->context, 0);
+
+        pthread_mutex_lock(&pool->lock);
+        if (!--pool->busy)
+            pthread_cond_signal(&pool->done);
+    }
+    pthread_mutex_unlock(&pool->lock);
+    return NULL;
+}
+
+// Starts the helpers from the rendering thread, whose priority they inherit.
+static void start_helpers(RenderPool *pool)
+{
+    pool->started = true;
+    int n = pool->renderer->settings.threads - 1;
+    pool->helpers = calloc(n, sizeof(*pool->helpers));
+    if (!pool->helpers)
+        return;
+
+    for (int i = 0; i < n; i++) {
+        Helper *helper = pool->helpers + i;
+        helper->pool = pool;
+        if (!ass_render_context_init(&helper->context, pool->renderer)) {
+            ass_render_context_done(&helper->context);
+            break;
+        }
+        if (pthread_create(&helper->thread, NULL, run_helper, helper)) {
+            ass_render_context_done(&helper->context);
+            break;
+        }
+#if defined(__linux__)
+        pthread_setname_np(helper->thread, "libass/render");
+#endif
+        pool->n_helpers++;
+    }
+}
+
+// Wakes helpers until, with the calling thread, that many work on the frame.
+static void wake_helpers(RenderPool *pool, int threads)
+{
+    if (threads < 2)
+        return;
+    if (!pool->started)
+        start_helpers(pool);
+
+    pthread_mutex_lock(&pool->lock);
+    int n = FFMIN(threads - 1, pool->n_helpers) - pool->busy - pool->wanted;
+    if (n > 0)
+        pool->wanted += n;
+    pthread_mutex_unlock(&pool->lock);
+    for (int i = 0; i < n; i++)
+        pthread_cond_signal(&pool->work);
+}
+
+static void wait_for_helpers(RenderPool *pool)
+{
+    if (!pool->n_helpers)
+        return;
+
+    pthread_mutex_lock(&pool->lock);
+    pool->wanted = 0;
+    while (pool->busy)
+        pthread_cond_wait(&pool->done, &pool->lock);
+    pthread_mutex_unlock(&pool->lock);
+
+    for (int i = 0; i < pool->n_helpers; i++)
+        ass_cache_flush(&pool->helpers[i].context.cache_client);
+}
+
+static bool reserve_events(RenderPool *pool, int count)
+{
+    if (count <= pool->max_events)
+        return true;
+
+    int size = FFMAX(count, 2 * pool->max_events);
+    PlannedEvent *plan = realloc(pool->plan, size * sizeof(*plan));
+    if (plan)
+        pool->plan = plan;
+    bool *fresh = realloc(pool->fresh, size * sizeof(*fresh));
+    if (fresh)
+        pool->fresh = fresh;
+    EventCost *costs = realloc(pool->costs, size * sizeof(*costs));
+    if (costs)
+        pool->costs = costs;
+    if (!plan || !fresh || !costs)
+        return false;
+
+    pool->max_events = size;
+    return true;
+}
+
+// The estimate for new events like this one, or the unused slot for it
+static NewEventCost *find_new_event_cost(RenderPool *pool, ASS_Event *event)
+{
+    unsigned slot = ((unsigned) event->Layer * 2654435761u ^ event->Style) & (NEW_COST_SLOTS - 1);
+    for (int n = 0; n < NEW_COST_SLOTS; n++, slot = (slot + 1) & (NEW_COST_SLOTS - 1)) {
+        NewEventCost *entry = pool->new_costs + slot;
+        if (!entry->used || (entry->layer == event->Layer && entry->style == event->Style))
+            return entry;
+    }
+    return NULL;
+}
+
+// Expects each event's cost, through cost_us.
+static void predict_costs(RenderPool *pool, int count)
+{
+    ASS_Track *track = pool->renderer->track;
+    EventImages *eimg = pool->renderer->eimg;
+    if (pool->costs_track != track) {
+        pool->n_costs = 0;
+        memset(pool->new_costs, 0, sizeof(pool->new_costs));
+    }
+
+    for (int i = 0, j = 0; i < count; i++) {
+        ASS_Event *event = eimg[i].event;
+        int index = event - track->events;
+        while (j < pool->n_costs && pool->costs[j].index < index)
+            j++;
+        pool->fresh[i] = j == pool->n_costs || pool->costs[j].index != index ||
+                         pool->costs[j].read_order != event->ReadOrder;
+        if (!pool->fresh[i]) {
+            eimg[i].cost_us = pool->costs[j].cost_us;
+            continue;
+        }
+        NewEventCost *estimate = find_new_event_cost(pool, event);
+        eimg[i].cost_us = estimate && estimate->used ? estimate->cost_us : NEW_EVENT_US;
+    }
+}
+
+static int compare_cost(const void *a, const void *b)
+{
+    uint32_t ca = ((const PlannedEvent *) a)->cost_us, cb = ((const PlannedEvent *) b)->cost_us;
+    return (ca < cb) - (ca > cb);
+}
+
+// Orders the frame's events, the heavy ones first and costliest first, and
+// returns how many threads to render them with.
+static int plan_frame(RenderPool *pool, int count)
+{
+    EventImages *eimg = pool->renderer->eimg;
+    predict_costs(pool, count);
+
+    int64_t total = 0;
+    uint32_t max_cost = 0;
+    int n_heavy = 0;
+    for (int i = 0; i < count; i++) {
+        uint32_t cost = eimg[i].cost_us;
+        total += cost;
+        max_cost = FFMAX(max_cost, cost);
+        if (cost >= HEAVY_EVENT_US)
+            pool->plan[n_heavy++] = (PlannedEvent) { cost, i };
+    }
+    qsort(pool->plan, n_heavy, sizeof(*pool->plan), compare_cost);
+    for (int i = 0, n = n_heavy; i < count; i++)
+        if (eimg[i].cost_us < HEAVY_EVENT_US)
+            pool->plan[n++] = (PlannedEvent) { eimg[i].cost_us, i };
+
+    // A frame can't take less than its costliest event.
+    int64_t share = FFMAX(max_cost, TARGET_US);
+    int threads = FFMINMAX((total + share - 1) / share, 1, pool->max_threads);
+
+    pool->count = count;
+    pool->n_heavy = n_heavy;
+    pool->batch = FFMINMAX((count - n_heavy) / (8 * threads), 1, 32);
+    atomic_store_explicit(&pool->next, 0, memory_order_relaxed);
+    return threads;
+}
+
+static int64_t take_wait(CacheClient *client)
+{
+    int64_t wait_us = client->wait_us;
+    client->wait_us = 0;
+    return wait_us;
+}
+
+static void adapt_threads(RenderPool *pool, int count)
+{
+    int64_t work = 0, waited = take_wait(&pool->renderer->state.cache_client);
+    for (int i = 0; i < count; i++)
+        work += pool->renderer->eimg[i].cost_us;
+    for (int i = 0; i < pool->n_helpers; i++)
+        waited += take_wait(&pool->helpers[i].context.cache_client);
+    int used = pool->joined + 1;
+    pool->joined = 0;
+
+    if (used > 1 && waited * 4 > work) {
+        pool->calm = 0;
+        if (++pool->contended >= 2) {
+            pool->max_threads = FFMAX(used / 2, 1);
+            pool->contended = 0;
+        }
+        return;
+    }
+    pool->contended = 0;
+    if (pool->max_threads < pool->renderer->settings.threads && ++pool->calm >= CALM_FRAMES) {
+        pool->max_threads++;
+        pool->calm = 0;
+    }
+}
+
+static void save_costs(RenderPool *pool, int count)
+{
+    ASS_Track *track = pool->renderer->track;
+    EventImages *eimg = pool->renderer->eimg;
+    for (int i = 0; i < count; i++) {
+        ASS_Event *event = eimg[i].event;
+        uint32_t cost = eimg[i].cost_us;
+        pool->costs[i] = (EventCost) { event - track->events, event->ReadOrder, cost };
+        if (!pool->fresh[i])
+            continue;
+        NewEventCost *estimate = find_new_event_cost(pool, event);
+        if (!estimate)
+            continue;
+        if (estimate->used)
+            estimate->cost_us += ((int64_t) cost - estimate->cost_us) / 4;
+        else
+            *estimate = (NewEventCost) { true, event->Layer, event->Style, cost };
+    }
+    pool->n_costs = count;
+    pool->costs_track = track;
+}
+
+static RenderPool *get_pool(ASS_Renderer *priv)
+{
+    if (!priv->pool && priv->settings.threads > 1) {
+        RenderPool *pool = calloc(1, sizeof(*pool));
+        if (!pool)
+            return NULL;
+        pool->renderer = priv;
+        pool->max_threads = priv->settings.threads;
+        pthread_mutex_init(&pool->lock, NULL);
+        pthread_cond_init(&pool->work, NULL);
+        pthread_cond_init(&pool->done, NULL);
+        priv->pool = pool;
+    }
+    return priv->pool;
+}
+
+void ass_render_events(ASS_Renderer *priv, int count)
+{
+    RenderPool *pool = get_pool(priv);
+    if (!pool || !reserve_events(pool, count)) {
+        for (int i = 0; i < count; i++) {
+            EventImages *event_images = priv->eimg + i;
+            event_images->rendered = ass_render_event(&priv->state, event_images->event,
+                                                      event_images);
+        }
+        return;
+    }
+
+    int threads = plan_frame(pool, count);
+    wake_helpers(pool, threads);
+    bool all_awake = pool->started && threads > pool->n_helpers;
+    run_queue(pool, &priv->state, all_awake ? 0 : ass_now_us() + TARGET_US);
+    wait_for_helpers(pool);
+    adapt_threads(pool, count);
+    save_costs(pool, count);
+}
+
+void ass_render_pool_free(RenderPool *pool)
+{
+    if (!pool)
+        return;
+
+    pthread_mutex_lock(&pool->lock);
+    pool->quit = true;
+    pthread_mutex_unlock(&pool->lock);
+    pthread_cond_broadcast(&pool->work);
+    for (int i = 0; i < pool->n_helpers; i++) {
+        pthread_join(pool->helpers[i].thread, NULL);
+        ass_render_context_done(&pool->helpers[i].context);
+    }
+
+    pthread_cond_destroy(&pool->done);
+    pthread_cond_destroy(&pool->work);
+    pthread_mutex_destroy(&pool->lock);
+    free(pool->helpers);
+    free(pool->plan);
+    free(pool->fresh);
+    free(pool->costs);
+    free(pool);
+}
diff --git a/libass/ass_utils.h b/libass/ass_utils.h
index 976b801..9bb9285 100644
--- a/libass/ass_utils.h
+++ b/libass/ass_utils.h
@@ -36,6 +36,7 @@
 
 #include "ass.h"
 #include "ass_string.h"
+#include <time.h>
 
 #ifndef SIZE_MAX
 #define SIZE_MAX ((size_t)-1)
@@ -255,6 +256,13 @@ static inline int mystrtod(char **p, double *res)
     return *p != start;
 }
 
+static inline int64_t ass_now_us(void)
+{
+    struct timespec ts;
+    clock_gettime(CLOCK_MONOTONIC, &ts);
+    return (int64_t) ts.tv_sec * 1000000 + ts.tv_nsec / 1000;
+}
+
 // strtoll clamped to int32, for bases 10 and 16
 static inline int mystrtoi32(char **p, int base, int32_t *res)
 {
diff --git a/libass/libass.sym b/libass/libass.sym
index 4cc6215..0430260 100644
--- a/libass/libass.sym
+++ b/libass/libass.sym
@@ -36,6 +36,7 @@ ass_process_force_style
 ass_set_message_cb
 ass_fonts_update
 ass_set_cache_limits
+ass_set_threads
 ass_flush_events
 ass_set_shaper
 ass_set_line_position
diff --git a/libass/meson.build b/libass/meson.build
index 9eda958..0d5d501 100644
--- a/libass/meson.build
+++ b/libass/meson.build
@@ -62,6 +62,7 @@ libass_src = files(
     'ass_rasterizer.c',
     'ass_render.c',
     'ass_render_api.c',
+    'ass_render_pool.c',
     'ass_shaper.c',
     'ass_string.c',
     'ass_strtod.c',
-- 
2.55.0.windows.5

